### ISSUE TITLE
[NuGet.VisualStudioExtension] Updating a package on a solution with 26 projects takes crazy long

### ISSUE THREAD

**Issue by GeertvanHorrik** [NuGet/Home#547]
Average CPU usage during update: 70 %
Memory usage: 2GB +

![image](https://cloud.githubusercontent.com/assets/1246444/7526081/14bdbbbe-f50f-11e4-82cd-8254057153a7.png)

Regular on this sln is just a few percentages and max 500 mb of memory.

Package Manager Console Host Version 3.0.0.0

Type 'get-help NuGet' to see all available NuGet commands.

```
Unable to satisfy package dependency constraints.
PM> update-package Catel.Core -IncludePrerelease
Attempting to gather dependencies information for package 'Catel.Core.4.2.0-unstable0077' with respect to project targeting '.NETFramework, Version=v4.0'
Attempting to resolve dependencies for package 'Catel.Core.4.2.0-unstable0077' with DependencyBehavior 'Lowest'
Resolving actions to install package 'Catel.Core.4.2.0-unstable0077'
Resolved actions to install package 'Catel.Core.4.2.0-unstable0077'
Removed package 'Catel.Core.4.2.0-unstable0074' from 'packages.config'
Successfully uninstalled 'Catel.Core.4.2.0-unstable0074' from Gum.Ui.NET40.
Adding package 'Catel.Core 4.2.0-unstable0077' to folder 'C:\Source\Gum.Ui\src\..\lib'
Added package 'Catel.Core 4.2.0-unstable0077' to folder 'C:\Source\Gum.Ui\src\..\lib'
Added package 'Catel.Core 4.2.0-unstable0077' to 'packages.config'
Successfully installed 'Catel.Core 4.2.0-unstable0077' to Gum.Ui.NET40.
Attempting to gather dependencies information for package 'Catel.Core.4.2.0-unstable0077' with respect to project targeting '.NETFramework, Version=v4.0'
Attempting to resolve dependencies for package 'Catel.Core.4.2.0-unstable0077' with DependencyBehavior 'Lowest'
Resolving actions to install package 'Catel.Core.4.2.0-unstable0077'
Resolved actions to install package 'Catel.Core.4.2.0-unstable0077'
Removed package 'Catel.Core.4.2.0-unstable0074' from 'packages.config'
Successfully uninstalled 'Catel.Core.4.2.0-unstable0074' from Gum.Ui.Apps.NET40.
Package 'Catel.Core 4.2.0-unstable0077' already exists in folder 'C:\Source\Gum.Ui\src\..\lib'
Added package 'Catel.Core 4.2.0-unstable0077' to 'packages.config'
Successfully installed 'Catel.Core 4.2.0-unstable0077' to Gum.Ui.Apps.NET40.
Attempting to gather dependencies information for package 'Catel.Core.4.2.0-unstable0077' with respect to project targeting '.NETFramework, Version=v4.0'
```

At first I tried to invoke via the UI => unresponsive. Then I restarted VS and used the powershell command: 

```
Update-Package Catel.Core -IncludePrerelease
```

Additional info:

* This project uses Shared projects to share code between different implementations.
* I have ReSharper 9.1 installed

Let me know if you need more info.

**Comment by deepakaravindr** [comment]
@GeertvanHorrik 
We have a few quick questions

1. Did you try without Resharper installed?
2. Can you share the packages.config from the various projects?
3. List of (enabled) package sources?

**Comment by GeertvanHorrik** [comment]
1) I have disabled resharper now
2) I have removed the v3 package source

Seems to be much faster now. It's still "slow", but it's the same as v2. Memory usage stays at 500 mb. More info soon.

**Comment by GeertvanHorrik** [comment]
Without resharper, with api.nuget.org package source: very slow, CPU jumps to 40 % (minimal, 70 % most of the time), memory goes up like crazy.

Below are the details of the package source:

1) api.nuget.org => https://api.nuget.org/v3/index.json
2) nuget.org => https://www.nuget.org/api/v2/
3) myget => https://mycompany.myget.org/F/myfeed/

Machine wide:

4) Microsoft and .NET =>  https://www.nuget.org/api/v2/curated-feeds/microsoftdotnet/

I think the cause is the v3 api.nuget.org

**Comment by GeertvanHorrik** [comment]
btw. when I open package manager console:

```
Each package is licensed to you by its owner. Microsoft is not responsible for, nor does it grant any licenses to, third-party packages. Some packages may include dependencies which are governed by additional licenses. Follow the package source (feed) URL to determine any dependencies.

Package Manager Console Host Version 3.0.0.0

Type 'get-help NuGet' to see all available NuGet commands.

Unable to satisfy package dependency constraints.
PM> 
```

Is the "Unable to satisfy package dependency constraints" normal?

**Comment by deepakaravindr** [comment]
in V3 we fixed a bug where we used to stop walking the sources if we found a potential match. Now we look through all of them in parallel and if you have a slow source (i'm guessing myget is the one) you are limited by the slowest source speed.

Running on V3 side by side with V2 works fine, however V2 is much slower than V3 so it's advisable to only have api.nuget.org turned on. We have just committed a fix to allow for them to be side by side with the same name so older versions of visual studio will keep working while Visual Studio 14 will pick up the newer faster feed.

At the same time, the memory usage seems alarming. Can you share a project where we can try this out and improve the performance?

For the powershell comment - This looks like some of the packages are not restored, and the error message is indeed confusing. We are going to open a separate issue for this. The true meaning of the message is that we cannot run init.ps1 scripts on some of the packages because they are missing.

**Comment by GeertvanHorrik** [comment]
Switched to api.nuget.org and tried to update a package (Catel.Core) on one of our projects via the console (which should only use a single package source ,right?).

This step takes a very long time (at least 20 seconds). Note that the package has no dependencies at all. And this for all projects in the solution (this time 12), so updating takes way too long and CPU usage goes up like crazy.

```
Attempting to gather dependencies information for package 'Catel.Core.4.2.0-unstable0078' with respect to project targeting '.NETFramework, Version=v4.0'
```

I think the issue is inside the gather dependency step. Between the rest of the updates, CPU goes down and works good, except for the dependency determination (and memory goes up to at least 1200 MB on a 12 project solution).

**Comment by GeertvanHorrik** [comment]
Just created a new WPF project => search for Orchestra.Shell.Ribbon.Fluent (including prerelease) => install takes crazy long, cpu goes up to 40 %, memory already at 1600 mb.

**Comment by johnataylor** [comment]
This should be fix in the next drop where we have a bunch of fixes to the resolver.
